jablib/src/main/java/org/jabref/logic/externalfiles/LinkedFileTransferHelper.java

```diff
@@ -0,0 +1,212 @@
+package org.jabref.logic.externalfiles;
+
+import java.io.IOException;
+import java.nio.file.FileAlreadyExistsException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Optional;
+import java.util.Set;
+
+import org.jabref.logic.FilePreferences;
+import org.jabref.logic.util.io.FileUtil;
+import org.jabref.model.database.BibDatabaseContext;
+import org.jabref.model.entry.BibEntry;
+import org.jabref.model.entry.LinkedFile;
+
+import org.jspecify.annotations.NonNull;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class LinkedFileTransferHelper {
+
+    private record FileCopyContext(
+      BibDatabaseContext sourceContext,
+      BibDatabaseContext targetContext,
+      FilePreferences filePreferences
+    ) { }
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(LinkedFileTransferHelper.class);
+
+    /**
+     * Adjusts linked files when copying entries from source to target context.
+     * Files that are not reachable from the target context will be copied.
+     * Files in the target context whose relative paths differ from the source will have their paths adjusted
+     *
+     * @param sourceContext The source database context where files are currently located
+     * @param targetContext The target database context where files should be accessible
+     * @param filePreferences File preferences for both contexts
+     */
+    public static Set<BibEntry> adjustLinkedFilesForTarget(
+        BibDatabaseContext sourceContext,
+        BibDatabaseContext targetContext,
+        FilePreferences filePreferences
+    ) {
+        if (!filePreferences.shouldAdjustOrCopyLinkedFilesOnTransfer()) {
+            return Set.of();
+        }
+
+        Set<BibEntry> modifiedEntries = new HashSet<>();
+
+        FileCopyContext context = new FileCopyContext(sourceContext, targetContext, filePreferences);
+
+        for (BibEntry entry : targetContext.getEntries()) {
+            boolean entryChanged = false;
+            List<LinkedFile> linkedFiles = new ArrayList<>();
+
+            for (LinkedFile linkedFile : entry.getFiles()) {
+                if (linkedFile.getLink().isEmpty()) {
+                  linkedFiles.add(linkedFile);
+                  continue;
+                }
+
+                Optional<Path> sourcePathOpt = linkedFile.findIn(sourceContext, filePreferences);
+                Optional<Path> targetPrimaryOpt = getPrimaryPath(targetContext, filePreferences);
+
+                if (sourcePathOpt.isEmpty() || targetPrimaryOpt.isEmpty()) {
```

Comment:
Add comment above

    // In case the file wasn't found at the source or the target - just re-use the existing link

Reply:
I don't understand this `if`. Why is the thing stopping if the fiel is not found in the targetPrimaryPath? There is no copy made before (is it?)